### DISCUSSION TITLE
Implement some form of "performance" report

### DISCUSSION THREAD

**Discussion by andybee** [architect/architect#1255] [Feature Requests]
There's lots of common tips and improvements when working with Architect, like trying to keep functions as small as possible to speed up deployment and execution.

It would be great if you could pass a flag to deploy (or a more appropriate tool?) that gives you a snapshot of your project, perhaps listing each function that's been considered for deployment and profile it's size.

What I'm really after is an easy way to glance and how big things like `shared` and `views` have got as I'm working on a codebase and to spot obvious outliers (why is that function so big, etc.)

**Comment by ryanblock** [comment]
I like this idea. A few questions:
- Performance can a bit tricky to quantify; we have some best practices, but what do you think we should specifically be looking at and calling out?
  - Memory size? Runtime version? Code payload size? (What else?)
  - Example of limitations of a performance report: you may have a very lean `3008MB` Lambda that's only 10KB so it coldstarts fast, but if you call bcrypt with a ridiculous number of rounds, it's still going to take forever to execute
- When and how should this snapshot be presented?

**Comment by andybee** [comment]
Regarding what to track, I'm not certain on that one beyond my specific use case (final size of the function once hydrated). I intentionally described it wider in my original request so that it could include other metrics that people might consider important. Some of your suggestions seem reasonable, though.

I feel like this will be an intensive task to run consistently, so feels like some kind of flag to be run as part of a dummy deploy? e.g. that's the point where we'd be getting everything ready as would be deployed, so best chance to run tests. Then in terms of output, the obvious thing is a CLI output - but I guess the concern is that for larger projects that will become hard to parse so maybe also/alternatively some form of disc output (log file? HTML summary?)

I wonder if there's existing libraries that could be brought in to do some of that report heavy lifting so we can move to just writing the measurement logic.
